sklearn/metrics/pairwise.py

```diff
@@ -244,6 +246,51 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     else:
         YY = row_norms(Y, squared=True)[np.newaxis, :]
 
+    if XX.size > 0 and YY.size > 0:
+        # Heuristics to detect possible numerical precision issues, with
+        # the used quadratic expansion that occurs when the distance between
+        # considered vectors is much lower than their norm.
+        #
+        # Here, a warning is raised if all considered vectors are within a ND
+        # sperical shell with a thickness to diameter ratio lower than
+        # the threshold value of 1e-7 in 64 bit. While such a configuration
+        # is generally uncommon, it occurs when a distribution with an
+        # inertia (or mean radius) of I, is shifted by I/threshold out of
+        # of the origin.
+
+        # min and max of sample norms, ignoring outliers
+        XX_min = nanpercentile(XX, 5)**0.5
+        XX_max = nanpercentile(XX, 95)**0.5
+        YY_min = nanpercentile(YY, 5)**0.5
+        YY_max = nanpercentile(YY, 95)**0.5
+        XY_min = min(XX_min, YY_min)
+        XY_max = max(XX_max, YY_max)
+
+        # only float64, float32 dtypes are possible
+        if X.dtype == np.float64:
+            threshold = 1e-7
+        elif X.dtype == np.float32:
+            threshold = 1e-3
+
+        if (abs(XY_max - XY_min) < threshold*XY_max and
+                # ignore null vector
+                XX_max > 0.0 and YY_max > 0
+                # ignore comparison between the vector and itself
+                and XX_max != YY_max):
+
+            warning_message = (
+                    "with the provided data, computing "
+                    "Euclidean distances with the quadratic expansion may "
+                    "lead to numerically inaccurate results. ")
+            if not issparse(X):
+                warning_message += (
+                    "Consider standardizing features by removing the mean, "
+                    "or setting globally "
+                    "euclidean_distances_algorithm='exact' for slower but "
```

Comment:
should we not just back off to using the exact algorithm in this case?

Reply:
Standardizing, centering etc. will not help in all cases. Consider the 1d data set -10001,-10000,10000,10001 with FP32; essentially the symmetric version of the example I provided before. The mean is zero; and scaling it will not resolve the relative closeness of the pairs of values.

Reply:
I agree, that sample statistics is not the solution here. This was merely an attempt at prototyping. Theoretically though since we already have the norms, we could use comparison between the norms as a proxy to determine potentially problematic points at the cost of `O(n_samples_a*n_samples_b)` as compared to the overall cost `O(n_samples_a*n_samples_b*n_features)` for the full calculation. Or less if we use some tree structure. Then recompute those points exactly, though it would have it's limitations...